src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSTree.java

```diff
@@ -0,0 +1,322 @@
+        return tree.get(id).getName();
+    }
+
+    public String getParentOf(final String id) {
+        Utils.validateArg(tree.containsKey(id), "Could not get parent of node id " + id + " because it does not exist");
+        return tree.get(id).getParent();
+    }
+
+    public long getLengthOf(final String id) {
+        Utils.validateArg(tree.containsKey(id), "Could not get length of node id " + id + " because it does not exist");
+        return tree.get(id).getLength();
+    }
+
+    public String getRankOf(final String id) {
+        Utils.validateArg(tree.containsKey(id), "Could not get rank of node id " + id + " because it does not exist");
+        return tree.get(id).getRank();
+    }
+
+    public boolean hasNode(final String id) {
+        return tree.containsKey(id);
+    }
+
+    /**
+     * Removes all nodes not in the given set
+     */
+    public void retainNodes(final Set<String> idsToKeep) {
+        Utils.validateArg(idsToKeep.contains(root), "Cannot remove root");
+        final HashMap<String, PSTreeNode> new_tree = new HashMap<>(idsToKeep.size());
+        for (final String id : tree.keySet()) {
+            if (idsToKeep.contains(id)) {
+                final PSTreeNode info = tree.get(id);
+                final PSTreeNode new_info = info.copy();
+                for (final String child : info.getChildren()) {
+                    if (!idsToKeep.contains(child)) {
+                        new_info.removeChild(child);
+                    }
+                }
+                if (!idsToKeep.contains(info.getParent())) {
+                    new_info.setParent(null);
+                }
+                new_tree.put(id, new_info);
+            }
+        }
+        tree = new_tree;
+    }
+
+    public void setNameOf(final String id, final String name) {
+        Utils.validateArg(tree.containsKey(id), "Could not set name of node id " + id + " because it does not exist");
+        Utils.validateArg(name != null, "Cannot set node name to null");
+        Utils.validateArg(!id.equals(root), "Cannot set name of root");
+        tree.get(id).setName(name);
+    }
+
+    public void setRankOf(final String id, final String rank) {
+        Utils.validateArg(tree.containsKey(id), "Could not set rank of node id " + id + " because it does not exist");
+        Utils.validateArg(rank != null, "Cannot set node rank to null");
+        Utils.validateArg(!id.equals(root), "Cannot set rank of root");
+        tree.get(id).setRank(rank);
+    }
+
+    public void setLengthOf(final String id, final long length) {
+        Utils.validateArg(tree.containsKey(id), "Could not set rank of node id " + id + " because it does not exist");
+        Utils.validateArg(!id.equals(root), "Cannot set rank of root");
+        tree.get(id).setLength(length);
+    }
+
+    /**
+     * Returns path of node id's from the input id to the root.
+     */
+    public List<String> getPathOf(String id) {
```

Comment:
As I mentioned before, it looks like this will infinite loop if the tree has a cycle. Perhaps you want to check for that an throw an exception, or at least document why if you don't think that'll happen in practice.